compiler/rustc_codegen_llvm/src/abi.rs

```diff
@@ -415,9 +415,31 @@ impl<'ll, 'tcx> FnAbiLlvmExt<'ll, 'tcx> for FnAbi<'tcx, Ty<'tcx>> {
             i += 1;
             i - 1
         };
+
+        let apply_range_attr = |idx: AttributePlace, abi: rustc_target::abi::Abi| {
+            if cx.sess().opts.optimize != config::OptLevel::No
+                && llvm_util::get_version() >= (19, 0, 0)
+                && let abi::Abi::Scalar(scalar) = abi
+                && matches!(scalar.primitive(), Int(..))
+                // If the value is a boolean, the range is 0..2 and that ultimately
+                // become 0..0 when the type becomes i1, which would be rejected
+                // by the LLVM verifier.
+                && !scalar.is_bool()
+                // LLVM also rejects full range.
+                && !scalar.is_always_valid(cx)
+            {
+                attributes::apply_to_llfn(
+                    llfn,
+                    idx,
+                    &[llvm::CreateRangeAttr(cx.llcx, scalar.size(cx), scalar.valid_range(cx))],
+                );
+            }
+        };
```

Comment:
...that this isn't called in enough places. Instead of building the attribute here, I would suggest:

- add a `valid_range` field to `struct ArgAttributes`
- set `valid_range` in `adjust_for_rust_scalar` when applicable
- call `llvm::CreateRangeAttr` in `rustc_codegen_llvm::abi::get_attrs`, based on the value of `valid_range` 

Then the existing code should handle the rest.

Reply:
This solution seems to use a lot more memory if I add a valid_range_and_num_bits: Option<(WrappingRange, u32)> to ArgAttributes the ArgAbi size goes from 56 to 192 and  FnAbi from 80 to 224 and the WrappingRange will not be interned but the layout is so even the same types will result in same ranges manytimes.

Reply:
updated scalarPair also to avoid undoing the work from https://github.com/rust-lang/rust/pull/100999 but there maybe is some good reason to do like you suggested like other backends want to use the data also.

Reply:
ended up at this assert and did not know of a good reason to increase it.
https://github.com/andjo403/rust/blob/3cbfa8212a48f8eecce9c5da0615da6c4c883ce9/compiler/rustc_target/src/abi/call/mod.rs#L970-L980